### PR TITLE
Local IPs detection

### LINKED ISSUE #1661: Detect localhost address and automatically appends to API server cert SANs
Currently we hardcode both `localhost` and `127.0.0.1` as SANs into the cert we generate for API server. But in some cases (e.g. Hetzner cloud) the boxes are configured with localhost pointing to `127.0.1.1`

We should try to detect to which local IP localhost is actually pointing at and add that to SANs automatically.

The current workaround is to add these to `spec.api.sans` list manually.

### PR DESCRIPTION
**Pull request by makhov**
Signed-off-by: makhov <amakhov@mirantis.com>

## Description

Local IPs detection to append them to the cert SANs. Currently, we use `127.0.0.1` which is not always the case.

Fixes #1661. 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

- [x] Manual test
- [ ] Auto test added

## Checklist:

- [x] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [x] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by jnummelin** [comment]
I don't think this cover the Hetzner specific case as the address `127.0.1.1` is NOT actually configured on the interfaces:
```
root@k0s-ubuntu20:~# hostname -i
127.0.1.1
root@k0s-ubuntu20:~# ip add show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 96:00:01:3f:4d:32 brd ff:ff:ff:ff:ff:ff
    inet 167.235.251.210/32 scope global dynamic eth0
       valid_lft 68554sec preferred_lft 68554sec
    inet6 2a01:4f8:1c17:cf11::1/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::9400:1ff:fe3f:4d32/64 scope link 
       valid_lft forever preferred_lft forever
```

Maybe we should do the `hostname -i` equivalent in the code to detect this case?

File: cmd/controller/certificates.go [hunk 1]
```diff
@@ -232,6 +237,36 @@ func (c *Certificates) Init(ctx context.Context) error {
...
+	}
+	for _, i := range ifaces {
+		addrs, err := i.Addrs()
+		if err != nil {
+			return nil, err
+		}
+
+		for _, addr := range addrs {
+			switch v := addr.(type) {
+			case *net.IPNet:
```
Comment (twz123) [inline cmd/controller/certificates.go:254]: Not sure if we want whole CIDRs?

File: cmd/controller/certificates.go [hunk 2]
```diff
@@ -232,6 +237,36 @@ func (c *Certificates) Init(ctx context.Context) error {
...
+				if v.IP.To4() == nil {
+					continue
+				}
+				localIPs = append(localIPs, v.IP.To4().String())
+			case *net.IPAddr:
+				if v.IP.To4() == nil {
+					continue
+				}
+				localIPs = append(localIPs, v.IP.To4().String())
```
Comment (twz123) [inline cmd/controller/certificates.go:263]: Isn't the linked issue just about localhost? Maybe we should filter for everything in the 127.0.0.0/8 subnet here?

**Comment by twz123** [comment]
> Maybe we should do the `hostname -i` equivalent in the code to detect this case?

For the record: it [uses](https://sourceforge.net/p/net-tools/code/ci/v2.10/tree/hostname.c#l155) `gethostbyname` and [prints](https://sourceforge.net/p/net-tools/code/ci/v2.10/tree/hostname.c#l194) the `h_addr_list` from the `hostent` struct.

Closest in Golang is probably `net.LookupHost` for both `"localhost"` and `os.Hostname()`.

File: cmd/controller/certificates.go (hunk 1 above)
Comment (makhov) [reply cmd/controller/certificates.go:254]: We don't, we add only IPs

File: cmd/controller/certificates.go (hunk 2 above)
Comment (makhov) [reply cmd/controller/certificates.go:263]: Right, will fix

**Comment by makhov** [comment]
`hostname -i` on my AWS instance returns `172.31.28.213`, and the same IP for `host $(hostname)`
I think `net.LookupHost("localhost")` will be a good option

File: cmd/controller/certificates.go [hunk 3]
```diff
@@ -232,6 +237,20 @@ func (c *Certificates) Init(ctx context.Context) error {
 	return eg.Wait()
 }
 
+func detectLocalIP() (string, error) {
+	addrs, err := net.LookupIP("localhost")
```
Comment (twz123) [inline cmd/controller/certificates.go:241]: I suspect that we need to do it twice, once for localhost, and once for the hostname itself? On my box, localhost resolves to 127.0.0.1, and my hostname resolves to 127.0.0.2

File: cmd/controller/certificates.go [hunk 4]
```diff
@@ -189,10 +190,14 @@ func (c *Certificates) Init(ctx context.Context) error {
 		"kubernetes.default.svc",
 		"kubernetes.default.svc.cluster",
 		fmt.Sprintf("kubernetes.svc.%s", c.ClusterSpec.Network.ClusterDomain),
-		"127.0.0.1",
```
Comment (twz123) [inline cmd/controller/certificates.go:192]: Might be worthwhile to include 127.0.0.1 in any case, even if localhost and the hostname resolve to something different.

**Review by twz123 — DISMISSED** [review]

File: cmd/controller/certificates.go (hunk 4 above)
Comment (makhov) [reply cmd/controller/certificates.go:192]: Done

**Review by twz123 — APPROVED** [review]

**Pull request merged**